Puzzle.java

```diff
@@ -0,0 +1,328 @@
+        closedList = new HashSet<>();
+
+    }
+
+    // method sot get the user input for the start state of the puzzle
+    public void userInput() {
+
+        Scanner scanner = new Scanner(System.in);
+        currentState = new int[3][3]; // initalize the 3x3 board
+
+        System.out.println("Enter the start state in a (3x3 grid, seperated with space from numbers 0-8)");
+        System.out.println("Example:\n2 3 4\n1 5 6\n8 7 0");
+        System.out.println();
+
+        // read the user input to generate the start state
+        for (int i = 0; i < 3; i++) {
+            for (int j = 0; j < 3; j++) {
+
+                currentState[i][j] = scanner.nextInt();
+
+            }
+        }
+
+        System.out.println("Enter '1' to use the Manhattan Distance or '2' to use Misplaced tiles heuristic ");
+        int choice = scanner.nextInt();
+        useManhattan = (choice == 1); // if the input is '1' then use manhattan
+
+    }
+
+    // A* search to solve the puzzle
+    public void solve() {
+
+        // create the start state with g = 0 and heuristic value (h(n))
+        State startState = new State(currentState, 0, calculateSumOfError(currentState));
+        openList.add(startState); // add start state to the openList
+        nodesGenerated++;
+
+        // loop until a solution is found or no more states are available to explore
+        while (!openList.isEmpty()) {
+
+            State current = openList.poll(); // get the current state that has the lowest cost (g + h)
+            nodesExpanded++;
+            closedList.add(current); // mark the state as explored
+
+            // want to check if the current state is the actual goal state
+            if (Arrays.deepEquals(current.board, goalState)) {
+
+                printSolution(current); // print the solution path
+                return;
+
+            }
+
+            // find and explore the neighboring states
+            makeAMove(current);
+
+        }
+
+        // at the end if no solution was found print
+        System.out.println("No Solution found");
+
+    }
+
+    // get all possible moves from the current state
+    private void makeAMove(State current) {
+
+        List<State> neighbors = new ArrayList<>();
+
+        // this is for trying to move in all four of the directions and find the
+        // neighboting states
+        neighbors.addAll(moveLeft(current));
+        neighbors.addAll(moveRight(current));
+        neighbors.addAll(moveUp(current));
+        neighbors.addAll(moveDown(current));
+
+        // add the neighbors that are valid to the open list if they havent been
+        // explored yet
+        for (State neighbor : neighbors) {
+
+            if (!closedList.contains(neighbor)) {
+                openList.add(neighbor);
+                nodesGenerated++;
+            }
+
+        }
+
+        // after getting the neighbors, add them to the closed List
```

Comment:
i don't need this, it's already doing that previously, if i add this it'll ruin the program